src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/LanguageServerTarget.cs

```diff
@@ -0,0 +1,427 @@
+
+    /// <inheritdoc cref="IRequestHandler{RequestContextType}.RequiresLSPSolution" />
+    bool RequiresLSPSolution { get; }
+
+    /// <inheritdoc cref="IRequestHandler{RequestContextType}.MutatesSolutionState" />
+    bool MutatesSolutionState { get; }
+
+    string MethodName { get; }
+
+    /// <summary>
+    /// The document identifier that will be used to find the solution and document for this request. This comes from the <see cref="TextDocumentIdentifier"/> returned from the handler itself via a call to <see cref="IRequestHandler{RequestType, ResponseType, TResponseContextType}.GetTextDocumentIdentifier(RequestType)"/>.
+    /// </summary>
+    TextDocumentIdentifier? TextDocument { get; }
+
+    /// <inheritdoc cref="RequestContext.ClientCapabilities" />
+    ClientCapabilities ClientCapabilities { get; }
+
+    /// <summary>
+    /// <see cref="CorrelationManager.ActivityId"/> used to properly correlate this work with the loghub
+    /// tracing/logging subsystem.
+    /// </summary>
+    Guid ActivityId { get; }
+
+    IRequestMetrics Metrics { get; }
+}
+
+public interface IRequestMetrics
+{
+    void RecordCancellation();
+    void RecordExecutionStart();
+    void RecordFailure();
+    void RecordSuccess();
+}
+
+public record RequestHandlerMetadata(string MethodName, Type RequestType, Type ResponseType);
+
+public interface IRequestDispatcher<RequestContextType> where RequestContextType : struct
+{
+    ImmutableArray<RequestHandlerMetadata> GetRegisteredMethods();
+
+    Task<TResponseType?> ExecuteRequestAsync<TRequestType, TResponseType>(
+        string methodName,
+        TRequestType request,
+        ClientCapabilities clientCapabilities,
+        IRequestExecutionQueue<RequestContextType> queue,
+        CancellationToken cancellationToken);
+}
+
+public abstract class LanguageServerTarget<RequestContextType> : ILanguageServer where RequestContextType : struct
+{
+    private readonly ICapabilitiesProvider _capabilitiesProvider;
+
+    protected readonly JsonRpc _jsonRpc;
+    private readonly IRequestDispatcher<RequestContextType> _requestDispatcher;
+    private readonly IRequestExecutionQueue<RequestContextType> _queue;
+    protected readonly ILspServices _lspServices;
+    protected readonly ILspLogger _logger;
+
+    // Set on first LSP initialize request.
+    protected readonly IClientCapabilitiesProvider _clientCapabilitiesProvider;
+
+    // Fields used during shutdown.
+    private bool _shuttingDown;
+    private Task? _errorShutdownTask;
+
+    public bool HasShutdownStarted => _shuttingDown;
+
+    public abstract InitializeParams ClientSettings { get; }
+
+    public event EventHandler<bool>? Shutdown;
+
+    public event EventHandler? Exit;
+
+    protected LanguageServerTarget(
+        ILspServiceProvider lspServiceProvider,
+        JsonRpc jsonRpc,
+        ICapabilitiesProvider capabilitiesProvider,
+        ILspLogger logger,
+        ImmutableArray<string> supportedLanguages,
```

Comment:
hmm, is this tailored towards Roslyn for C#/VB things?

Reply:
Good catch, that needs to be stripped out since it's Roslyn specific.